awx/main/dispatch/worker/callback.py

```diff
@@ -164,38 +163,46 @@ def flush(self, force=False):
                     else:  # only calculate the seconds if the created time already has been set
                         metrics_total_job_event_processing_seconds += e.modified - e.created
                 metrics_duration_to_save = time.perf_counter()
+                saved_events = []
                 try:
                     cls.objects.bulk_create(events)
                     metrics_bulk_events_saved += len(events)
+                    saved_events = events
+                    self.buff[cls] = []
                 except Exception as exc:
-                    logger.warning(f'Error in events bulk_create, will try indiviually up to 5 errors, error {str(exc)}')
+                    # if the database is flaking, let ensure_connection throw a general exception
+                    # will be caught by the outer loop, which goes into a proper sleep and retry loop
+                    django_connection.ensure_connection()  # INCLUDE in final patch
+                    logger.warning(f'Error in events bulk_create, will try indiviually, error: {str(exc)}')
                     # if an exception occurs, we should re-attempt to save the
                     # events one-by-one, because something in the list is
                     # broken/stale
-                    consecutive_errors = 0
-                    events_saved = 0
                     metrics_events_batch_save_errors += 1
-                    for e in events:
+                    for e in events.copy():
+                        # special sanitization logic for postgres treatment of NUL 0x00 char
+                        if "\x00" in e.stdout:
```

Comment:
Kind of nervous about the performance implications of scanning every single byte of stdout here. Would it be better to catch ValueError, check if it contains "A string literal cannot contain NUL", and then fallback to doing this?

Reply:
This is already in the exception block for `bulk_create`. So this only triggers when an error happens, but it runs against all events, including the perfectly good ones. Event errors (outside of database problems) _should_ be extremely rare, but things don't always work as they should.

To be more honest about things - the handling was put earlier than it should have been to make the unit test possible.

To follow the logic, this is where I see it going:

```diff
diff --git a/awx/main/dispatch/worker/callback.py b/awx/main/dispatch/worker/callback.py
index 2bb0699d35..9d6f988af7 100644
--- a/awx/main/dispatch/worker/callback.py
+++ b/awx/main/dispatch/worker/callback.py
@@ -179,10 +179,6 @@ class CallbackBrokerWorker(BaseWorker):
                     # broken/stale
                     metrics_events_batch_save_errors += 1
                     for e in events.copy():
-                        # special sanitization logic for postgres treatment of NUL 0x00 char
-                        if "\x00" in e.stdout:
-                            e.stdout = e.stdout.replace("\x00", "")
-
                         try:
                             e.save()
                             metrics_singular_events_saved += 1
@@ -191,6 +187,11 @@ class CallbackBrokerWorker(BaseWorker):
                         except Exception as exc_indv:
                             retry_count = getattr(e, '_retry_count', 0) + 1
                             e._retry_count = retry_count
+
+                            # special sanitization logic for postgres treatment of NUL 0x00 char
+                            if (retry_count == 1) and isinstance(exc_indv, ValueError) and ("\x00" in e.stdout):
+                                e.stdout = e.stdout.replace("\x00", "")
+
                             logger.info(f'Database Error Saving individual Event uuid={e.uuid} try={retry_count}, error: {str(exc_indv)}')
                             if retry_count >= self.INDIVIDUAL_EVENT_RETRIES:
                                 logger.error(f'Hit max retries saving individual Event error: {str(exc_indv)}\ndata:\n{e.__dict__}')
```

I don't have any argument against this. But doing this, I would give up on the corresponding test, because I know this would get too complicated to realistically maintain.

Reply:
Pushed a commit where I did this but kept the test, adding a couple of mocks instead. The test isn't quite as bad as I feared, so we'll go with it.

Reply:
Thanks for clarifying Alan.